_posts/2017-05-10-polyglot-scientific-computing-with-xtensor.md

```diff
@@ -0,0 +1,195 @@
+
+## Enforcing the usage of expressions
+
+Our template `logsumexp` function is generic, however it allows to use any type for the first argument. If we try to call it with something that does not implement the `xexpression` interface, we will end up with cryptic errors, the compiler complaining about not being able to convert the parameter type to a very complicated template type.
+
+To avoid this, we can modify `logsumexp` signature so it complains earlier when its first parameter type is not an expression:
+
+```cpp
+template <class E>
+E logsumexp(const xt::xexpression<E>& e, const axes_type& axes)
+{
+    const E& a = e.derived_cast();
+    auto m = xt::eval(xt::amax(a, axes));
+    auto sv = xt::slice_vector(m);
+    for (int i = 0; i < e1.dimension(); i++)
+    {
+        if (std::find (axes.begin(), axes.end(),  i) != axes.end())
+            sv.push_back(xt::newaxis());
+        else
+            sv.push_back(xt::all());
+    }
+    auto max2 = xt::dynamic_view(m, sv);
+    return E(m + xt::log(xt::sum(xt::exp(a - max2), axes)));
+}
+```
+
+The first line in the function retrieves the real type of the expression, so that we can keep the rest of the implementation unchanged. Now you can see the benefits of wrapping `logsumexp` inside `python_logsumexp`: we don't have to change anything else in the code. Whereas if we had used direct instantiation of the template, the following changes would have been required:
+
+```cpp
+PYBIND11_PLUGIN(fast_lse)
+{
+    xt::import_numpy();
+    py::module m("fast_lse", "Fast implementation of logsumexp);
+    // m.def("logsumexp", logsumexp<array_type>);
+    m.def("logsumexp", logsumexp<xt::xexpression<array_type>>);
+    return m.ptr();
+}
+```
+
+## Splitting the project
+
+Now that the implementation of `logsumexp` is generic enough, we can move it to a dedicated header. That is not sufficient, though. Indeed, even if including this header in a pure C++ module does not make that module depend on numpy, it makes any package including your C++ module depend on both `xtensor-python` and `pybind11`.
+
+We can avoid useless dependencies by splitting the projects:
+- A pure C++ project containing the template function `logsumexp`.
+- A python extension project depending on the previous one, and exposing the `logsumexp` function to the Python.
+
+This way, we can use `logsumexp` from C++ modules without any dependency on `pybind11` or `xtensor-python`. And since `logsumexp` is generic, we can use it with other kinds of expressions, such as `xtensor`, `xarray`, database backends, file backends, etc ...
+
+But we can go further. We can now expose `logsumexp` to Julia with very few additional work:
+
+```cpp
+#include <cxx_wrap.hpp>
+#include "xtensor-julia/jlarray.hpp
+#include "fast_lse/logsumexp.hpp"
+
+using array_type = xt::jlarray<double>;
+using axes_type = cxx_wrap::ArrayRef<std::size_t>
```

Comment:
The signed int type of julia if typedeffed to `int_t`. Maybe we should use that?